src/ProbeBase.jl

```diff
@@ -2,138 +2,337 @@ module ProbeBase
+                Threads.atomic_sub!(spec.semaphore, 1)
+            end
+        end
     end
-    Threads.atomic_sub!(spec.semaphore, 1)
 end
-function enable_probes(mod::Module)
-    lock(PROBES_LOCK) do
-        for spec in values(PROBES[mod])
+function enable!(mod::Module, category::Symbol)
+    mod_lock = mod.__probebase_lock__
+    tracepoints = mod.__probebase_tracepoints__
+    lock(mod_lock) do
+        for spec in values(tracepoints[category])
             Threads.atomic_add!(spec.semaphore, 1)
         end
     end
 end
-function disable_probes(mod::Module)
-    lock(PROBES_LOCK) do
-        for spec in values(PROBES[mod])
+function disable!(mod::Module, category::Symbol)
+    mod_lock = mod.__probebase_lock__
+    tracepoints = mod.__probebase_tracepoints__
+    lock(mod_lock) do
+        for spec in values(tracepoints[category])
             Threads.atomic_sub!(spec.semaphore, 1)
         end
     end
 end
-
-probe_enabled(spec::ProbeSpec) = spec.semaphore[] > 0
-@generated function probe_trigger(spec::ProbeSpec, args...)
-    Targs = Expr(:tuple, map(nameof, args)...)
-    ex = Expr(:call, :ccall, :(spec.payload[]), :Cvoid, Targs)
-    for (idx, arg) in enumerate(args)
-        push!(ex.args, :(args[$idx]))
+function enable!(mod::Module, category::Symbol, kind::Symbol)
+    mod_lock = mod.__probebase_lock__
+    tracepoints = mod.__probebase_tracepoints__
+    lock(mod_lock) do
+        Threads.atomic_add!(tracepoints[category][kind].semaphore, 1)
+    end
+end
+function disable!(mod::Module, category::Symbol, kind::Symbol)
+    mod_lock = mod.__probebase_lock__
+    tracepoints = mod.__probebase_tracepoints__
+    lock(mod_lock) do
+        Threads.atomic_sub!(tracepoints[category][kind].semaphore, 1)
     end
-    ex
 end
-function probe_maybe_trigger(spec::ProbeSpec, kind::Symbol, args...)
+
+probe_enabled(spec::TracepointSpec) = spec.semaphore[] > 0
+function probe_maybe_trigger(spec::TracepointSpec, category::Symbol, kind::Symbol, lib_id::Int64, args...)
     if probe_enabled(spec)
-        probe_trigger(spec, args...)
+        probe_trigger(spec, category, kind, lib_id, args...)
     end
 end
+@generated function probe_trigger(spec::TracepointSpec, category::Symbol, kind::Symbol, lib_id::Int64, arg)
+    Targs = Expr(:tuple, :Symbol, :Symbol, :Int64, :Any)
+    ex = Expr(:call, :ccall, :(spec.payload[]), :Int64, Targs)
+    push!(ex.args, :(category))
+    push!(ex.args, :(kind))
+    push!(ex.args, :(lib_id))
+    push!(ex.args, :(arg))
+    return ex
+end
+
+function register_probe(mod::Module, category::Symbol, kind::Symbol, spec::TracepointSpec; enable=false)
+    if enable
+        spec.semaphore[] = 1
+    end
+    if !isdefined(mod, :__probebase_lock__)
+        mod_lock = mod.eval(:(__probebase_lock__ = Threads.ReentrantLock()))
+        tracepoints = mod.eval(:(__probebase_tracepoints__ = Dict{Symbol,Dict{Symbol,$TracepointSpec}}()))
+        funcs = mod.eval(:(__probebase_funcs__ = Dict{Symbol,Dict{Symbol,Any}}()))
```

Comment:
Is `__probebase_funcs__` just to provide GC rooting?

Reply:
Yep! I should probably document that.